protocol/messenger_communities.go

```diff
@@ -2732,220 +2721,296 @@ func (m *Messenger) RequestImportDiscordCommunity(request *requests.ImportDiscor
-			importProgress.StopTask(discord.DownloadAssetsTask)
-			progressUpdates <- importProgress
-			cancel <- communityID
-			return
-		}
-
-		// Signal to the client that save operations are starting
-		importProgress.UpdateTaskState(discord.DownloadAssetsTask, discord.TaskStateSaving)
-		progressUpdates <- importProgress
+					assetCounter.Increase()
+					progressValue := calculateProgress(i+1, totalImportChunkCount, (float32(assetCounter.Value())/float32(totalAssetsCount))*0.25)
+					importProgress.UpdateTaskProgress(discord.DownloadAssetsTask, progressValue)
+					progressUpdates <- importProgress
 
-		// We chunk message attachments by `maxChunkSizeBytes` to ensure individual
-		// save operations don't take too long and block the database
-		attachmentChunks := chunkAttachmentsByByteSize(messageAttachmentsToDownload, maxChunkSizeBytes)
-		chunksCount = len(attachmentChunks)
+				}(id, author)
+			}
+			wg.Wait()
 
-		for i, attachments := range attachmentChunks {
-			err = m.persistence.SaveDiscordMessageAttachments(attachments)
-			if err != nil {
-				m.cleanUpImport(communityID)
-				importProgress.AddTaskError(discord.DownloadAssetsTask, discord.Error(err.Error()))
-				importProgress.Stop()
+			if m.DiscordImportMarkedAsCancelled(communityID) {
+				importProgress.StopTask(discord.DownloadAssetsTask)
 				progressUpdates <- importProgress
+				cancel <- communityID
 				return
 			}
 
+			for idxRange := range gopart.Partition(len(messageAttachmentsToDownload), 100) {
+				attachments := messageAttachmentsToDownload[idxRange.Low:idxRange.High]
+				wg.Add(1)
+				go func(attachments []*protobuf.DiscordMessageAttachment) {
+					defer wg.Done()
+					for ii, attachment := range attachments {
+
+						m.communitiesManager.LogStdout(fmt.Sprintf("downloading asset %d/%d", assetCounter.Value()+1, totalAssetsCount))
+
+						assetPayload, contentType, err := discord.DownloadAsset(attachment.Url)
+						if err != nil {
+							errmsg := fmt.Sprintf("Couldn't download message attachment '%s': %s", attachment.Url, err.Error())
+							importProgress.AddTaskError(
+								discord.DownloadAssetsTask,
+								discord.Warning(errmsg),
+							)
+							progressUpdates <- importProgress
+							continue
+						}
+
+						attachment.Payload = assetPayload
+						attachment.ContentType = contentType
+						messageAttachmentsToDownload[ii] = attachment
+
+						if m.DiscordImportMarkedAsCancelled(communityID) {
+							importProgress.StopTask(discord.DownloadAssetsTask)
+							progressUpdates <- importProgress
+							cancel <- communityID
+							return
+						}
+
+						assetCounter.Increase()
+						progressValue := calculateProgress(i+1, totalImportChunkCount, 0.25+(float32(assetCounter.Value())/float32(totalAssetsCount))*0.25)
+						importProgress.UpdateTaskProgress(discord.DownloadAssetsTask, progressValue)
+						progressUpdates <- importProgress
+					}
+				}(attachments)
+			}
+			wg.Wait()
```

Comment:
While this looks like a new thing, it's just been moved around as we now put the import process into a loop (on a per file basis)